src/chainstate/coordinator/mod.rs

```diff
@@ -682,6 +682,128 @@ fn forget_orphan_stacks_blocks(
     Ok(())
 }
 
+/// Get the heaviest affirmation map, when considering epochs.
+/// * In epoch 2.05 and prior, the heaviest AM was the sortition AM.
+/// * In epoch 2.1, the reward cycles prior to the 2.1 boundary remain the sortition AM.
+pub fn static_get_heaviest_affirmation_map(
+    burnchain: &Burnchain,
+    burnchain_blocks_db: &BurnchainDB,
+    sortition_db: &SortitionDB,
+    sortition_tip: &SortitionId,
+) -> Result<AffirmationMap, Error> {
+    let last_2_05_rc = sortition_db.get_last_epoch_2_05_reward_cycle()? as usize;
+
+    let sort_am = sortition_db.find_sortition_tip_affirmation_map(sortition_tip)?;
+
+    let heaviest_am = BurnchainDB::get_heaviest_anchor_block_affirmation_map(
+        burnchain_blocks_db.conn(),
+        burnchain,
+    )?;
+
+    let mut am_entries = vec![];
+    for i in 0..last_2_05_rc {
+        if i < sort_am.affirmations.len() {
+            am_entries.push(sort_am.affirmations[i]);
+        } else {
+            return Ok(AffirmationMap::new(am_entries));
+        }
+    }
+    for i in last_2_05_rc..heaviest_am.len() {
+        am_entries.push(heaviest_am.affirmations[i]);
+    }
+
+    Ok(AffirmationMap::new(am_entries))
+}
+
+/// Get the canonical affirmation map, when considering epochs.
+/// * In epoch 2.05 and prior, the heaviest AM was the sortition AM.
+/// * In epoch 2.1, the reward cycles prior to the 2.1 boundary remain the sortition AM.
+pub fn static_get_canonical_affirmation_map(
+    burnchain: &Burnchain,
+    burnchain_blocks_db: &BurnchainDB,
+    sortition_db: &SortitionDB,
+    chain_state_db: &StacksChainState,
+    sortition_tip: &SortitionId,
+) -> Result<AffirmationMap, Error> {
+    let last_2_05_rc = sortition_db.get_last_epoch_2_05_reward_cycle()? as usize;
+
+    let sort_am = sortition_db.find_sortition_tip_affirmation_map(sortition_tip)?;
+
+    let canonical_am = StacksChainState::find_canonical_affirmation_map(
+        burnchain,
+        burnchain_blocks_db,
+        chain_state_db,
+    )?;
+
+    let mut am_entries = vec![];
+    for i in 0..last_2_05_rc {
+        if i < sort_am.affirmations.len() {
+            am_entries.push(sort_am.affirmations[i]);
+        } else {
+            return Ok(AffirmationMap::new(am_entries));
+        }
+    }
+    for i in last_2_05_rc..canonical_am.len() {
+        am_entries.push(canonical_am.affirmations[i]);
+    }
+
+    Ok(AffirmationMap::new(am_entries))
+}
+
+fn inner_static_get_stacks_tip_affirmation_map(
+    burnchain_blocks_db: &BurnchainDB,
+    last_2_05_rc: u64,
+    sort_am: &AffirmationMap,
+    sortdb_conn: &DBConn,
+    canonical_ch: &ConsensusHash,
+    canonical_bhh: &BlockHeaderHash,
+) -> Result<AffirmationMap, Error> {
+    let last_2_05_rc = last_2_05_rc as usize;
```

Comment:
the algorithm looks the same (or very similar) to `static_get_canonical_affirmation_map`. Should we make a common function?

Reply:
It uses a different affirmation map (the canonical one, not the Stacks one).

I consolidated the code across all of these related methods to merge the affirmation maps with the sortition affirmation map for reward cycle affirmations in epoch 2.1.